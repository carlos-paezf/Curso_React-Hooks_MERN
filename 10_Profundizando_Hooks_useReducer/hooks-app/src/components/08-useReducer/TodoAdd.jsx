import React from 'react';
import { useForm } from '../../hooks';
import PropTypes from 'prop-types'


export const TodoAdd = ({ handleAddTodo }) => {
    const [{ desc }, handleInputChange, reset] = useForm({
        desc: ''
    })

    const handleSubmit = (e) => {
        e.preventDefault()
        if (desc.trim().length < 1) return
        const newTodo = {
            id: new Date().getTime(),
            desc: desc,
            done: false
        }
        handleAddTodo(newTodo)
        reset()
    }

    return (
        <>
            <h2>Agregar TODO</h2>
            <hr />

            <form onSubmit={handleSubmit} className="form-group">
                <input type="text" name="desc" value={desc} onChange={handleInputChange} className="form-control" placeholder="Aprender ..." />
                <button className="btn btn-outline-primary w-100 mt-3" type="submit">Agregar</button>
            </form>
        </>
    );
}; 


TodoAdd.propTypes = {
    handleAddTodo: PropTypes.func.isRequired
}