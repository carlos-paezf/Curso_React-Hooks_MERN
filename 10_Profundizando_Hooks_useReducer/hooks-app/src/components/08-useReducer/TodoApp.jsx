import React, { useReducer } from 'react';
import { todoReducer } from './todoReducer';


const initialState = [{
    id: new Date().getTime(),
    desc: 'Aprender React',
    done: false
}]


export const TodoApp = () => {

    const [todos] = useReducer(todoReducer, initialState)

    return (
        <>
            <h2>TodoApp ( {todos.length} )</h2>
            <hr />

            <div className="row">
                <div className="col-sm-12 col-md-8">
                    <ul className="list-group list-group-flush">
                        {todos.map(({ id, desc }, i) =>  (
                            <li key={i} className="list-group-item">
                                <p className="text-center">{i + 1}. {desc}</p>
                                <button className="btn btn-danger">Eliminar</button>
                            </li>
                        ))}
                    </ul>
                </div>

                <div className="col-sm-12 col-md-4">
                    <h2>Agregar TODO</h2>
                    <hr />

                    <form className="form-group">
                        <input className="form-control" type="text" name="desc" placeholder="Aprender ..." />
                        <button className="btn btn-outline-primary w-100 mt-3">Agregar</button>
                    </form>
                </div>
            </div>

        </>
    );
};

