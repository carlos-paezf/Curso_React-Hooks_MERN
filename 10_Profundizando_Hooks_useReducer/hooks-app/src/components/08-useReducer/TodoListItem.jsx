import React from 'react';
import PropTypes from 'prop-types'


export const TodoListItem = ({ i, todo, handleToggle, handleDelete }) => {
    const { id, desc, done } = todo

    return (
        <>
            <li key={i} className="list-group-item">
                <p onClick={() => handleToggle(id)} className={done ? "complete" : "text-center"}>{i + 1}. {desc}</p>
                <button onClick={() => handleDelete(id)} className="btn btn-danger">Eliminar</button>
            </li>
        </>
    );
};


TodoListItem.propTypes = {
    i: PropTypes.number.isRequired,
    todo: PropTypes.object.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}