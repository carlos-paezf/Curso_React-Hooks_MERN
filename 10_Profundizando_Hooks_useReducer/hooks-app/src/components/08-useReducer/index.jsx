export { TodoApp } from "./TodoApp";
export { TodoList } from "./TodoList";
export { TodoListItem } from "./TodoListItem";
export { TodoAdd } from "./TodoAdd";