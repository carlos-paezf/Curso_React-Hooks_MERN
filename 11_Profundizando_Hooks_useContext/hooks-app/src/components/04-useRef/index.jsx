export { FocusScreen } from './FocusScreen'
export { RealExampleRef } from './RealExampleRef'