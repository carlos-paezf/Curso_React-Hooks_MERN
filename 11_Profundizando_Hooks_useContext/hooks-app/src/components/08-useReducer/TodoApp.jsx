import React, { useEffect, useReducer } from 'react';
import { TodoAdd, TodoList } from '.';
import { todoReducer } from './todoReducer';


const init = () => JSON.parse(localStorage.getItem('todos')) || []


export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init)

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos))
    }, [todos])

    const handleDelete = (todoId) => { 
        dispatch({
            type: 'delete',
            payload: todoId
        })
    };

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    };

    const handleAddTodo = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        })
    };
    


    return (
        <>
            <h2>TodoApp ( {todos.length} )</h2>
            <hr />

            <div className="row">
                <div className="col-sm-12 col-md-8">
                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                </div>

                <div className="col-sm-12 col-md-4">
                    <TodoAdd handleAddTodo={handleAddTodo} />
                </div>
            </div>
        </>
    )
}
