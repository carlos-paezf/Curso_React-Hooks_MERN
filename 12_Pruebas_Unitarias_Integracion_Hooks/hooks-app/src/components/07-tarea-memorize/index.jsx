export { Hijo } from "./Hijo";
export { Padre } from "./Padre";