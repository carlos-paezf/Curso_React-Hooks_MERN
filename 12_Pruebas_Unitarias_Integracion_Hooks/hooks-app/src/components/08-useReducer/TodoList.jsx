import React from 'react';
import { TodoListItem } from '.';
import PropTypes from 'prop-types'


export const TodoList = ({ todos, handleToggle, handleDelete }) => {
    return (
        <>
            <ul className="list-group list-group-flush">
                {todos.map((todo, i) => <TodoListItem key={i} i={i} todo={todo} handleToggle={handleToggle} handleDelete={handleDelete} />)}
            </ul>
        </>
    );
};


TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleToggle: PropTypes.func.isRequired,
    handleDelete: PropTypes.func.isRequired
}