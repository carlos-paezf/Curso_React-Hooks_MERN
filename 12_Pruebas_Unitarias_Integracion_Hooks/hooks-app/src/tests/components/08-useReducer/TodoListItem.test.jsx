import React from 'react';
import { shallow } from 'enzyme'
import { TodoListItem } from '../../../components/08-useReducer';
import { demoTodos } from '../../fixtures/demoTodos';


/* The TodoListItem component is a stateless component that renders a single todo item.

The component has two props:

i: The index of the todo item.
todo: The todo object.

The component has two methods:

handleToggle: This method is called when the user clicks the checkbox.
handleDelete: This method is called when the user clicks the delete button. */
describe('Probar el componente <TodoListItem />', () => {
    let wrapper

    const handleToggle = jest.fn()
    const handleDelete = jest.fn()
    const i = 1

    beforeEach(() => {
        wrapper = shallow(
            <TodoListItem i={i} todo={demoTodos[1]} handleToggle={handleToggle} handleDelete={handleDelete} />
        )
    })

    test('Debe de mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('Debe de llamar la funciÃ³n borrar', () => {
        wrapper.find('button').simulate('click')
        expect(handleDelete).toHaveBeenCalled()
        expect(handleDelete).toHaveBeenCalledTimes(1)
        expect(handleDelete).toHaveBeenCalledWith(expect.any(Number))
        expect(handleDelete).toHaveBeenCalledWith(demoTodos[1].id)
    })

    test('Debe de mostrar el texto correctamente', () => {
        expect(wrapper.find('p').text().trim()).toBe(`${i + 1}. ${demoTodos[1].desc}`)
    })

    test('Debe de tener la clase complete si el TODO.done == true', () => {
        const todo = demoTodos[0]
        todo.done = true
        wrapper = shallow(
            <TodoListItem i={1} todo={todo} handleToggle={handleToggle} handleDelete={handleDelete} />
        )

        expect(wrapper.find('p').hasClass('complete')).toBe(true)
    })
})
