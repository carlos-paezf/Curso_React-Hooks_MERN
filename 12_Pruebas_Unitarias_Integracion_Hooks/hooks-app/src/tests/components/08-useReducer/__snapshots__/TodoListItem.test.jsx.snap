// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probar el componente <TodoListItem /> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      className="text-center"
      onClick={[Function]}
    >
      2
      . 
      Aprender React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Eliminar
    </button>
  </li>
</Fragment>
`;
